ext/standard/array.c

```diff
@@ -349,6 +350,102 @@ static zend_always_inline int php_array_data_compare_unstable_i(Bucket *f, Bucke
 }
 /* }}} */
 
+/* return int to be compatible with compare_func_t */
+static int hash_zval_strict_function(zval *z1, zval *z2) /* {{{ */
+{
+	ZVAL_DEREF(z1);
+	ZVAL_DEREF(z2);
+	// The most important thing about this comparison mode is that the result
+	// is 0 when zend_is_identical, and non-zero otherwise.
+	if (zend_is_identical(z1, z2)) {
+		return 0;
+	}
+	// Below this point, the return value should always be non-zero.
+	// If the types are different, compare based on type.
+	int t1 = Z_TYPE_P(z1);
+	int t2 = Z_TYPE_P(z2);
+	if ( t1 != t2 ) {
+		return (t1 > t2 ) ? 1 : -1;
+	}
+	// Both types are the same *but the values are not identical*
+	switch (t1) {
+	case IS_LONG:
+		return Z_LVAL_P(z1) > Z_LVAL_P(z2) ? 1 : -1;
+
+	case IS_DOUBLE:
+		return Z_DVAL_P(z1) > Z_DVAL_P(z2) ? 1 : -1;
+
+	case IS_STRING:
+		return zend_binary_strcmp(
+			Z_STRVAL_P(z1), Z_STRLEN_P(z1),
+			Z_STRVAL_P(z2), Z_STRLEN_P(z2)
+		);
+		
+	case IS_ARRAY:
+		// Do a recursive comparison.  Unlike for objects, if this recursive
+		// comparison returns 0, then the objects should have tested as
+		// identical above.
+		return zend_hash_compare(
+			Z_ARRVAL_P(z1), Z_ARRVAL_P(z2),
+			(compare_func_t) hash_zval_strict_function, 1 /* ordered */
+		);
+
+	case IS_OBJECT:
+	{
+		// Start with a recursive comparison like for arrays, for consistency.
```

Comment:
Why doesn't this only compare the object ID? That should be sufficient to establish an ordering.

Reply:
Because the object ID leads to nondeterministic behavior.  In my experience, having an array be sorted slightly differently every time you run the program is the sort of thing that leads to headaches.  Starting with a recursive comparison of attributes and properties is consistent with the way arrays are sorted, and the way that objects and arrays are compared with == (ie, that's the behavior of `zend_std_compare_objects`).  Comparing class names is an extra attempt to avoid non-determinism.  The object ID is only used as a fallback if your objects really are identical in all other ways.  Ie, the way I've implemented it here the order of the `==` elements will be nondeterministic, but all the other `!=` elements will have a consistent order from run to run.

Reply:
(in latest patch added test cases to document predictable order of non-identical objects, dropped the class name comparison.)

Reply:
(and to be clear by "nondeterminism" I mean that adding a new object allocation somewhere in your program, or reordering some code which performs allocations, can cause a nonlocal change in the sort order elsewhere.  I believe in PHP the object IDs are actually, strictly-speaking, deterministic given identical inputs to a program, but in practice in large programs it "feels" nondeterministic because of the effect of nonlocal changes.)

Reply:
Depends on what the goal here is. I don't think SORT_STRICT is particularly useful outside array_unique(). And array_unique() just needs some arbitrary order, so it should be the most efficient such order.

It may be worthwhile to accept SORT_STRICT in array_unique() only, to avoid exposing any particular strict ordering to userland -- it becomes an unobservable implementation detail.

Reply:
Yes, I started out thinking about just `array_unique`, but once you define it as a constant it seems to invite being used for all the different array methods which accept flags.  And I can think of use cases where a strict ordering is useful, for example anywhere where string and numeric values are mixed but you don't want the `"0"`, `"0.0"`, and `0` keys to all be considered the same. (`SORT_STRING` will conflate `"0"` and `0`, and `SORT_DEFAULT` and `SORT_NUMERIC` will conflate `"0"` and `"0.0"`).  Similarly if you want `null` to be a valid value sorted distinct from `""`.
"Sort, but don't do *any* conversion on the keys" actually seems like a genuinely-useful feature.
```
$a = [ "0", "0.0", "", null, "", "0.0", "0" ];
$flags = 0; // Nothing we can put here will put all identical values together
asort($a, $flags);
var_dump($a);
```

I'm not convinced that the sort order for `array_unique` is an "unobservable implementation detail" -- I think the actual sort order chosen is still revealed by the order of the keys in the returned array.  I have to think about that some more.